alea/include/alps/alea/core.hpp

```diff
@@ -217,49 +252,121 @@ struct reducer
     /** Set-up reduction operation */
     virtual reducer_setup get_setup() const = 0;
 
+    /** Get maximum of scalar value over all instances (immediate) */
+    virtual long get_max(long value) const = 0;
+
     /** Reduce double data-set into `data` */
-    virtual void reduce(sink<double> data) const = 0;
+    virtual void reduce(view<double> data) const = 0;
 
     /** Reduce long data-set into `data` */
-    virtual void reduce(sink<long> data) const = 0;
+    virtual void reduce(view<long> data) const = 0;
 
     /** Finish reduction of all data if deferred */
     virtual void commit() const = 0;
 
+    /** Returns a copy of `*this` created using `new` */
+    virtual reducer *clone() { throw unsupported_operation(); }
+
     /** Destructor */
     virtual ~reducer() { }
 
     // Convenience functions
 
-    void reduce(sink<std::complex<double> > data) const {
-        reduce(sink<double>((double *)data.data(), 2 * data.size()));
+    void reduce(view<std::complex<double> > data) const {
```

Comment:
what? why do you cast complex to pair of double? At least in case of mpi reduce you definitely shouldn't do this.

and why it's not reference?

Reply:
view is a wrapper pointer, so you don't need to pass it by reference.

Reply:
And why cannot I cast a complex to a pair of double? A complex literally *is* a pair of doubles, is it not?  And reduction of a complex is the reduction of real and imaginary part, is it not?

Reply:
Sure you can, but why? The only reason if your underlying library for reduction does not support complex, but at least mpi is not the case.